### PR TITLE
issue: 1006374 Tx path optimizations

### PR DESCRIPTION
**Pull request by liranoz12**
The following PR includes 8 different commits :

- Remove cq_mgr::poll_and_process_helper_rx() and cq_mgr::poll_and_process_helper_tx().
- Optimize qp_mgr::send().
- Add m_p_tx_mem_buf_desc_head to dst_entry which holds the next buffer to be sent.
  This change is only implemented for UDP.
- Remove m_rx_ready_byte_count counter from sockinfo.
   This variable is a duplicate of m_p_socket_stats->n_rx_ready_byte_count.
- 1. Convert block_this_run from variable into DEFINE.
  2. Remove is_dropped counter from sockinfo_udp::tx()
  3. Remove debug checks from dst_entry_tcp:fast_send()
  4. Postpone performing the following actions after send() completes:
      Ring migration checks, retransmission counters update and number of free wr update.
- Remove redundant variable "apifalgs" from tcp_write()
  This caus us about three redundant if statements in the fast path.
- Add pbuf_alloc to lwip tcp.h which holds the next pbuff to be sent
- Add seg_alloc to lwip tcp.h which holds the next tcp_segment to be sent.

Signed-off-by: Liran Oz <lirano@mellanox.com>

### CODE REVIEW

**Comment by liranoz12** [comment]
@DanielLibenson please review

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1411/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/942/ for details (Mellanox internal link).

**Comment by liranoz12** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/963/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1432/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
@liranoz12 please fix Jenkins

**Comment by OphirMunk** [comment]
@liranoz12 Please consider splitting different issues into different commits

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/978/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1448/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1449/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/979/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1455/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1477/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/993/ for details (Mellanox internal link).

File: src/vma/proto/dst_entry_tcp.cpp [hunk 1]
```diff
@@ -171,17 +159,11 @@ ssize_t dst_entry_tcp::fast_send(const struct iovec* p_iov, const ssize_t sz_iov
 		send_ring_buffer(m_id, m_p_send_wqe, b_blocked, is_dummy);
 	}
 
-#ifndef __COVERITY__
-        struct tcphdr* p_tcp_h = (struct tcphdr*)(((uint8_t*)(&(p_pkt->hdr.m_ip_hdr))+sizeof(p_pkt->hdr.m_ip_hdr)));
-        NOT_IN_USE(p_tcp_h); /* to supress warning in case VMA_OPTIMIZE_LOG */
-        dst_tcp_logfunc("Tx TCP segment info: src_port=%d, dst_port=%d, flags='%s%s%s%s%s%s' seq=%u, ack=%u, win=%u, payload_sz=%u",
```
Comment (rosenbaumalex) [inline src/vma/proto/dst_entry_tcp.cpp]: Better to keep the debug prints
I see they area already after the send operation. 
unless you do full debug prints cleanup for a RELEASE mode... very hard to debug cases later on

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/1539/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/1026/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
Merged #330

**Pull request closed without merging**